server/text_generation_server/utils/tokens.py

```diff
@@ -221,3 +149,159 @@ def from_pb(
             pb.max_new_tokens,
             pb.ignore_eos_token,
         )
+
+
+class HeterogeneousNextTokenChooser:
+    def __init__(
+        self,
+        dtype: torch.dtype,
+        device: torch.device,
+        watermark: List[bool],
+        temperature: List[float],
+        repetition_penalty: List[float],
+        top_k: List[int],
+        top_p: List[float],
+        typical_p: List[float],
+        do_sample: List[bool],
+        seeds: List[int],
+    ):
+        warpers = LogitsProcessorList()
+
+        if any(watermark):
+            raise NotImplementedError("Watermarking not implemented")
```

Comment:
Added support in my PR https://github.com/huggingface/text-generation-inference/pull/272/files#diff-412b58820dd5cecd65029639ebbe381fe19c58ca63faa94c246149a743c9da28R279

Reply:
Nice!